File: DesignPatterns/01_GOF23/DP_examples/Python/02_Structural/Composite/ConcreteComposite.py
from Component import Composite

# 员工类
class Employee(Composite):
    def __init__(self, is_base_employee):
        super().__init__(is_base_employee)
        self.name = ""
        self.sex = ""
        self.age = 0
        self.salary = 0
    
    # 初始化员工信息
    def init(self, name, sex, age, salary):
        self.name = name
        self.sex = sex
        self.age = age
        self.salary = salary
        return self
    
    # 重写 operation 方法
    def operation(self):
        print(f"[Name:{self.name}][Age:{self.age}][Sex:{self.sex}]:Doing some work, and ", end="")
        if self.sex == "man":
            print(f"his salary is {self.salary}.", end="")
        else:
            print(f"her salary is {self.salary}.", end="")
        if not self.is_leaf():
            print(f" who has {len(self._sub_composites)} Employees work under.")
        else:
            print()
        super().operation()

File: DesignPatterns/01_GOF23/DP_examples/Lua/02_Structural/Composite/ConcreteComposite.lua
require("Component")

-- 员工类
---@class Employee : Composite
---@field name string
---@field sex string
---@field age number
---@field salary number
Employee = Composite:new(false)

-- 构造函数
function Employee:new(isBaseEmployee)
    local o = Composite.new(self, isBaseEmployee)
    o.name = ""
    o.sex = ""
    o.age = 0
    o.salary = 0
    return o
end

-- 初始化员工信息
---@param name string
---@param sex string
---@param age number
---@param salary number
---@return Employee
function Employee:init(name, sex, age, salary)
    self.name = name
    self.sex = sex
    self.age = age
    self.salary = salary
    return self
end

function Employee:operation()
    io.write(string.format("[Name:%s][Age:%d][Sex:%s]:Doing some work, and ", self.name, self.age, self.sex))
    if self.sex == "man" then
        io.write(string.format("his salary is %d.", self.salary))
    else
        io.write(string.format("her salary is %d.", self.salary))
    end
    if not self.isLeaf then
        io.write(string.format(" who has %d Employees work under.", #self._subComposites))
    end
    io.write("\n")
    Composite.operation(self)
end
